lighthouse-core/lib/page-functions.js

```diff
@@ -113,18 +113,24 @@ function getElementsInDocument(selector) {
  */
 /* istanbul ignore next */
 function getOuterHTMLSnippet(element, ignoreAttrs = []) {
-  const clone = element.cloneNode();
-
-  ignoreAttrs.forEach(attribute =>{
-    clone.removeAttribute(attribute);
-  });
-
-  const reOpeningTag = /^[\s\S]*?>/;
-  const match = clone.outerHTML.match(reOpeningTag);
+  try {
+    if (ShadowRoot.prototype.isPrototypeOf(element) && element.host && element.localName !== 'a') {
```

Comment:
nit:
```suggestion
    if (element instanceof ShadowRoot && element.host && element.localName !== 'a') {
```
reads a little simpler than the prototype check, but

non-nit: can we also add a comment to this line? Both to explain the shadow root case, but also to explain the `localName` check. I don't really understand it...`ShadowRoot` doesn't have a `localName`, so by that point in the conditional won't the comparison always be true?

Reply:
(also according to spec ["Shadow roots’s associated `host` is never null."](https://dom.spec.whatwg.org/#interface-shadowroot))

Reply:
@patrickhulce since you added this line, what's the deal with the `element.localName !== 'a'` check?

Reply:
This entire piece is straight up copied from the dom-size audit that filters this way, so you'd have to ask Eric as to why it was written this way :)

It indeed seemed like overkill and I thought about trying to extract this to a `isShadowRoot` function with a canonical version to do this or something but bringing in all the baggage of `page-functions` with dependencies into this PR seemed unfair.

I don't see why we can't just use `element instanceof ShadowRoot` personally, but I'm assumed Eric thought of something I didn't which is a pretty safe bet when it comes to our comparative web component/shadow dom knowledge 😆

Reply:
:P I commited the suggested changes, 
is there any other issue that we need to tackle?

Reply:
> This entire piece is straight up copied from the dom-size audit that filters this way, so you'd have to ask Eric as to why it was written this way :)

OK! After consulting the old texts (#2131), I think this was from adapting earlier code (which identified shadow roots just by looking for an `element.host` but had to rule out anchors because they also have a `host` property). That code was wrapped without modification in the `ShadowRoot.prototype` check in that PR, which I believe was the error (but since it's always true, it doesn't cause any bad behavior).

I'm going to remove based on the spec and my own testing saying it's fine. I'll take responsibility for any failures :)